File: Contest/Weekly Contest 496/Integers With Multiple Sum of Two Cubes/Solution.cpp
// Problem Link : https://leetcode.com/problems/integers-with-multiple-sum-of-two-cubes/description/
using ll = long long;
class Solution {
public:
    vector<int> findGoodIntegers(int n) {
        int lim = cbrt(n);
        unordered_map<long long, int> mp;
        vector<int> ans;
        for (int a = 1; a <= lim; a++) {
            ll asum = 1LL * a * a * a;
            for (int b = a; b <= lim; b++) {
                ll bsum = 1LL * b * b * b;
                ll tsum = asum + bsum;
                if (tsum > n)
                    break;
                mp[tsum]++;
            }
        }
        for (auto& it : mp) {
            if (it.second >= 2)
                ans.push_back(it.first);
        }

        sort(ans.begin(), ans.end());
        return ans;
    }
};

File: Contest/Weekly Contest 496/Integers With Multiple Sum of Two Cubes/Solution.java
// Problem Link : https://leetcode.com/problems/integers-with-multiple-sum-of-two-cubes/description/
class Solution {
    public List<Integer> findGoodIntegers(int n) {
        Map<Long, Integer> mp = new HashMap<>();
        for (int i = 1; i <= 1000; i++) {
            long isum = (long) i * i * i;
            if (isum > n)
                break;
            for (int j = i; j <= 1000; j++) {
                long jsum = (long) j * j * j;
                long tsum = isum + jsum;
                if (tsum > n)
                    break;
                mp.put(tsum, mp.getOrDefault(tsum, 0) + 1);
            }
        }
        List<Integer> ans = new ArrayList<>();
        for (Map.Entry<Long, Integer> mpp : mp.entrySet()) {
            if (mpp.getValue() >= 2)
                ans.add(mpp.getKey().intValue());
        }
        Collections.sort(ans);
        return ans;
    }
}
